### PR TITLE
fix(workspaces): add pagination to workspace team members

### PR DESCRIPTION
**Pull request by cdriesler**
## Description & motivation

- We want to paginate queries to `workspace.team`

## Changes:

- Introduces pagination with the user name as the cursor and order
- Return `WorkspaceCollaboratorCollection` instead of a flat array from query
- Tests

### CODE REVIEW

File: packages/server/assets/workspacesCore/typedefs/workspaces.graphql
Comment (fabis94) [inline packages/server/assets/workspacesCore/typedefs/workspaces.graphql:148]: This will likely break Mike's members frontend, coordinate with him on how to release this. Maybe his adjustments to the FE should be done in this same branch to avoid deploying a BE change that breaks the FE.

File: packages/server/modules/workspaces/repositories/workspaces.ts
Comment (fabis94) [inline packages/server/modules/workspaces/repositories/workspaces.ts:171]: 1) There should be a default limit, otherwise the query is just gonna return everything in one long list
2) The limit, whatever it ends up being, should be clamped (lodash's `clamp()`) between an allowed minimum and maximum, lets say 0 and 100

File: packages/server/modules/workspaces/repositories/workspaces.ts
Comment (fabis94) [inline packages/server/modules/workspaces/repositories/workspaces.ts:197]: I don't think Users.name is a good cursor, there could be multiple people with the same name and that will likely break pagination. Also, I don't think you should order by Users.name, it makes more sense to show new collaborators first. Can we add a createdAt/updatedAt to workspaces_acl and order by/cursor by that?

If not, at least order by/cursor by User.createdAt (makes less sense tho).

Comment (fabis94) [reply packages/server/modules/workspaces/repositories/workspaces.ts:197]: Also cursors should be opaque: https://graphql.org/learn/pagination/

> Especially if the cursors are opaque, either offset or ID-based pagination can be implemented using cursor-based pagination (by making the cursor the offset or the ID), and using cursors gives additional flexibility if the pagination model changes in the future. As a reminder that the cursors are opaque and that their format should not be relied upon, we suggest base64 encoding them.

Use encode/decode cursor utilities from `import { decodeCursor, encodeCursor } from '@/modules/shared/helpers/graphqlHelper'`

File: packages/server/modules/workspaces/tests/integration/workspaces.graph.spec.ts
Comment (fabis94) [inline packages/server/modules/workspaces/tests/integration/workspaces.graph.spec.ts:202]: Add a test case for when you have multiple users with the same name

**Review by fabis94 — CHANGES_REQUESTED** [review]

**Comment by cdriesler** [comment]
Back to you @fabis94 , Mike is going to do the FE changes here before we merge. Welcome to wait for those or review backend changes now

File: packages/server/modules/workspaces/repositories/workspaces.ts
Comment (fabis94) [inline packages/server/modules/workspaces/repositories/workspaces.ts:180]: 1. Why do 'withoutTablePrefix' if you do actually add the table prefix, just manually?
2. You can use DbWorkspaceAcl.colAs() to alias columns
```
DbWorkspaceAcl.colAs('role', 'workspaceRole'),
DbWorkspaceAcl.colAs('createdAt', 'workspaceCreatedAt'),
DbWorkspaceAcl.colAs('updatedAt', 'workspaceUpdatedAt')
```

But that's all moot, cause you don't need to avoid aliases and you don't need to alias columns either - because the table aliases are already there. So if you do orderby() or groupby() DbWorkspaceAcl.col.createdAt, its already distinct from any other kind of createdAt because of the col prefix

File: packages/server/modules/workspaces/tests/integration/repositories.spec.ts
Comment (fabis94) [inline packages/server/modules/workspaces/tests/integration/repositories.spec.ts:147]: seems like there's a lot of duplication here, each of those upsertWorkspaceRole calls has only 1-2 different lines. that said - up to you to evaluate if they are fundamentally duplicate or only temporarily duplicate and probably gonna change a lot from each other in the future

File: packages/server/modules/workspaces/tests/integration/workspaces.graph.spec.ts
Comment (fabis94) [inline packages/server/modules/workspaces/tests/integration/workspaces.graph.spec.ts:190]: 50 seems too much, even wait(1) should be fine, I would think

File: packages/server/modules/workspaces/tests/unit/events/eventListener.spec.ts
Comment (fabis94) [inline packages/server/modules/workspaces/tests/unit/events/eventListener.spec.ts:18]: So many different places needing this duplicate block of createdAt/updatedAt being added makes me feel like the type should've just been changed instead to make createdAt/updatedAt optional and then possibly renamed to WorkspaceAclInsert to represent the ACL model pre-insertion, and then possibly having another type with those fields made required whenever you're returning them from the DB.

I dunno, might be too big of a change so maybe not worth doing, but having to duplicate these dates everywhere does feel icky

**Review by fabis94 — CHANGES_REQUESTED** [review]
nearly there, left some minor comments

File: packages/server/modules/workspaces/tests/integration/workspaces.graph.spec.ts
Comment (cdriesler) [reply packages/server/modules/workspaces/tests/integration/workspaces.graph.spec.ts]: +1 done

File: packages/server/modules/workspaces/repositories/workspaces.ts
Comment (cdriesler) [reply packages/server/modules/workspaces/repositories/workspaces.ts:182]: kind of silly, in retrospect. ty for breaking it down

Comment (fabis94) [reply packages/server/modules/workspaces/repositories/workspaces.ts:182]: I guess it may not be obvious, but DBHelper.col.X does add table prefixes by default, so you don't need to manually ever do that. Its only really for update calls where you want to use .withoutTablePrefix., cause `SET X = Y` instructions can't have table prefixes in them

**Comment by cdriesler** [comment]
Simplified query, deferring the WorkspaceAcl type change as discussed (but keeping in mind for future)

**Comment by cdriesler** [comment]
On pause pending frontend changes to use new gql query shape

**Comment by Mikehrn** [comment]
> On pause pending frontend changes to use new gql query shape

@cdriesler I started this Wednesday but wasn't able to finish it, will wrap it and up and push it today

File: packages/server/modules/workspacesCore/domain/types.ts
Comment (gjedlicska) [inline packages/server/modules/workspacesCore/domain/types.ts:20]: as discussed, we only need createdAt, updatedAt has not usefull meaning for now.

**Review by gjedlicska — CHANGES_REQUESTED** [review]

File: packages/server/modules/workspacesCore/domain/types.ts
Comment (cdriesler) [reply packages/server/modules/workspacesCore/domain/types.ts:24]: created the `workspace_acl_updates` table as an exercise - now we can keep `workspace_acl` unchanged

**Comment by cdriesler** [comment]
🚨 🚨 COOL SQL ALERT 🚨 🚨 

Created `workspace_acl_updates` table to log all changes to `workspace_acl` and give us some history.

Please give special attention to what I did in `getWorkspaceCollaboratorsFactory` to:

- join `workspace_acl_updates` (many) with `workspace_acl` (one)
- order things by the earliest `updatedAt` in `workspace_acl_updates` for user

**Comment by cdriesler** [comment]
DO NOT MERGE THIS

**Comment by cdriesler** [comment]
JK FALSE ALARM

**Review by gjedlicska — APPROVED** [review]

**Pull request merged**
